### PR TITLE
Refactor callbacks

### PR DESCRIPTION
**Pull request by kuynzereb**
Some refactoring of callbacks is done in this PR:
1. Added `on_validation_begin()` and `on_valdiation_end()` to callbacks. The point is that  `ModelCheckpoint` uses `on_epoch_end()` but actually is called in the validation end. It is fixed.
2. All callbacks calls are unified and take no additional arguments. That is, from now on it is `on_epoch_end()` instead of `on_epoch_end(epoch, logs)`. Instead of these additional arguments now callback will have a link to the trainer, so it will have access to the `current_epoch`, `global_step`, `callback_metrics` and so on. We just need to call `self.callback.set_trainer(self)` while initializing callbacks in the trainer.

With these modifications it will be easy to implement such things as additional checkpointing callback which uses `on_epoch_end` instead of `on_validation_end` and so can be used to checkpointing training with no validation loops (#596, #652). Also it will be easy to start using `global_step` in checkpoints name. And in general all callbacks will be more unified.

### CODE REVIEW

**Comment by williamFalcon** [comment]
@Borda @neggert good to go?

**Comment by Borda** [comment]
it is quite extensive change, ill have look tomorrow...

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 1]
```diff
@@ -30,40 +31,55 @@ def set_model(self, model):
             model = model.module
         self.model = model
 
-    def on_epoch_begin(self, epoch, logs=None):
+    def set_trainer(self, trainer):
+        self.trainer = trainer
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: add a docstring

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 2]
```diff
@@ -30,40 +31,55 @@ def set_model(self, model):
...
         called when the batch starts.
+        """
+        pass
 
-        Args:
-            batch (Tensor): current batch tensor
-            logs (dict): key-value pairs of quantities to monitor
+    def on_batch_end(self):
+        """
+        called when the batch ends.
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: when you use . at the end, make it as a sentence and it can be even a single line, so
```
"""Callback - when the batch ends."""

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 3]
```diff
@@ -149,13 +165,14 @@ def check_metrics(self, logs):
 
         return True
 
-    def on_train_begin(self, logs=None):
+    def on_train_begin(self):
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: just curious about the body here, this shall stay even after overwrite...

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 4]
```diff
@@ -149,13 +165,14 @@ def check_metrics(self, logs):
...
         # Allow instances to be re-used
         self.wait = 0
         self.stopped_epoch = 0
         self.best = np.Inf if self.monitor_op == np.less else -np.Inf
 
-    def on_epoch_end(self, epoch, logs=None):
+    def on_epoch_end(self):
+        logs = self.trainer.callback_metrics
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: pls add assert that the `self.trainer` is not `None`

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 5]
```diff
@@ -21,6 +21,7 @@ class Callback(object):
     def __init__(self):
         self.validation_data = None
         self.model = None
+        self.trainer = None
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: I would move all these as protected, so `self._model`, `self._trainer`, etc.

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 6]
```diff
@@ -167,13 +184,13 @@ def on_epoch_end(self, epoch, logs=None):
         else:
             self.wait += 1
             if self.wait >= self.patience:
-                self.stopped_epoch = epoch
+                self.stopped_epoch = self.trainer.current_epoch
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: all these class properties shall be listed in init, so it is clear what contains...

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 7]
```diff
@@ -30,40 +31,55 @@ def set_model(self, model):
             model = model.module
         self.model = model
 
-    def on_epoch_begin(self, epoch, logs=None):
+    def set_trainer(self, trainer):
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: as all callbacks depend in trainer now, would it make sense to pass trainer in init...

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 1 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Done.

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 2 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Done.

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 4 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Done.

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 5 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Done. Moreover, I have removed things like `.set_model()` and `.set_params()` because it was not really used anywhere.

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 6 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Hm, it seems like all these properties are already listed in init.

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 7 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Yeah, I thought about it. But it seems like a bad idea because when the user creates a callback there is no trainer at the moment. I mean, now the user should do the following:
```
callback = Callback(some_callback_params)
trainer = Trainer(callback=callback)
```
And there will be problems if trainer should be used as a callback parameter.

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 3 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Unfortunately, I don't understand what do you mean :)

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/callbacks/pt_callbacks.py]: I don't see `stopped_epoch` in init...

Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: It is right here https://github.com/PyTorchLightning/pytorch-lightning/blob/1b101d82747269ccf82480a4f22c441b3e44f28e/pytorch_lightning/callbacks/pt_callbacks.py#L105

**Comment by hadim** [comment]
I am planning to move the progress bar to a callback and I'll need this PR. Here are a few suggestions.

- could we have a `test_start` and `test_end`?
- could we move each callbacks in a separate file so maintenance and history are easier?
- Could we unify how callbacks are passed to the trainer? It would be useful to generalize and pass a list of callbacks instead of individual callbacks. During each event, the trainer should call all the callbacks. Free to the callback to add the necessary logic to do something or not.

Let me know what do you think.

**Comment by Borda** [comment]
> I am planning to move the progress bar to a callback and I'll need this PR. Here are a few suggestions.
> 
> * could we have a `test_start` and `test_end`?

Yes

> * could we move each callbacks in a separate file so maintenance and history are easier?

Good one :]

> * Could we unify how callbacks are passed to the trainer? It would be useful to generalize and pass a list of callbacks instead of individual callbacks. During each event, the trainer should call all the callbacks. Free to the callback to add the necessary logic to do something or not.

That would be nice...

> Let me know what do you think.

@williamFalcon ^^

**Comment by kuynzereb** [comment]
I can make first 2 suggestions. And I totally agree with the third suggestion and I thought about it myself. But I am afraid it is not very straightforward with the current implementation of callbacks and now I don't have enough time to delve into it. So I would like to implement only the first two suggestions in this PR.

**Comment by hadim** [comment]
That would be great. I make no promise but I can try to tackle the third one.

**Comment by Borda** [comment]
@kuynzereb @hadim I think that it would be much easier to pass if each suggestion will be single PR... the large/complex PRs are not so nice for review (takes much longer to check) nor the author (debugging may become quite complex) :]

**Comment by kuynzereb** [comment]
@Borda it also sounds very reasonable :)

**Comment by hadim** [comment]
No problem to make PRs but something like `test_start` and `test_end` could easily be added to this one without too much burden I think.

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 8]
```diff
@@ -12,60 +12,54 @@
...
+        """Called when the validation loop begins."""
         pass
 
-    def on_train_end(self, logs=None):
+    def on_validation_end(self):
+        """Called when the validation loop ends."""
         pass
 
 
+_no_trainer_error_msg = ".set_trainer() should be called after the callback initialization"
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]:
```
_NO_TRAINER_ERROR_MSG = "Missing trainer instance. The `.set_trainer(...)` should be called after the callback initialization."
```

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 9]
```diff
@@ -148,13 +142,16 @@ def check_metrics(self, logs):
...
-    def on_train_begin(self, logs=None):
+    def on_train_begin(self):
         # Allow instances to be re-used
         self.wait = 0
         self.stopped_epoch = 0
         self.best = np.Inf if self.monitor_op == np.less else -np.Inf
 
-    def on_epoch_end(self, epoch, logs=None):
+    def on_epoch_end(self):
+        assert self._trainer is not None, _no_trainer_error_msg
```
Comment (Borda) [inline pytorch_lightning/callbacks/pt_callbacks.py]: I would move this assert to a parent class and just at the beginning call parent method ?

File: tests/test_trainer.py [hunk 10]
```diff
@@ -231,8 +231,12 @@ def mock_save_function(filepath):
     # CASE K=-1  (all)
     w = ModelCheckpoint(save_dir, save_top_k=-1, verbose=1)
     w.save_function = mock_save_function
+    trainer = Trainer()
+    w.set_trainer(trainer)
```
Comment (Borda) [inline tests/test_trainer.py]: use longer var name then one letter

File: tests/test_trainer.py [hunk 11]
```diff
@@ -249,8 +253,12 @@ def mock_save_function(filepath):
     # CASE K=0 (none)
     w = ModelCheckpoint(save_dir, save_top_k=0, verbose=1)
     w.save_function = mock_save_function
+    trainer = Trainer()
+    w.set_trainer(trainer)
     for i, loss in enumerate(losses):
-        w.on_epoch_end(i, logs={'val_loss': loss})
+        w._trainer.current_epoch = i
+        w._trainer.callback_metrics = {'val_loss': loss}
```
Comment (Borda) [inline tests/test_trainer.py]: add a comment that this is kind of hack to simulate training...

**Comment by Borda** [comment]
@kuynzereb may you push a commit so it triggers new GitHub CI

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 8 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: Done

File: tests/test_trainer.py (hunk 10 above)
Comment (kuynzereb) [reply tests/test_trainer.py]: Done

File: tests/test_trainer.py (hunk 11 above)
Comment (kuynzereb) [reply tests/test_trainer.py]: Done

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 9 above)
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: I am not sure about this. Theoretically there may be some calls or even entire callbacks which don't use trainer at all. So there will be no need in this assert.

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 12]
```diff
@@ -12,60 +12,54 @@
 
 import numpy as np
 
-from pytorch_lightning.overrides.data_parallel import LightningDistributedDataParallel
-
 
 class Callback(object):
-    r"""Abstract base class used to build new callbacks.
-    """
+    """Abstract base class used to build new callbacks."""
```
[the remark this answers is not in the document]
Comment (kuynzereb) [reply pytorch_lightning/callbacks/pt_callbacks.py]: I am not familiar with all this `ABC` stuff, so I don't really know :)
But anyway I think it would be better to do it in another PR (if it really should be done)

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 9 above)
Comment (Borda) [reply pytorch_lightning/callbacks/pt_callbacks.py]: true... lets keep it for now

**Comment by Borda** [comment]
Great work, thx

**Comment by Borda** [comment]
it was nice to see that GH actions were about twice faster than Travis lol

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/callbacks/pt_callbacks.py]: fyi: https://www.python-course.eu/python3_abstract_classes.php

**Comment by williamFalcon** [comment]
@Borda can you close the appropriate tickets related to this PR?

**Comment by Borda** [comment]
I don't see any particular issue for this, but I will check the backlog later... @kuynzereb was this change requested in a issue?

**Comment by kuynzereb** [comment]
> @kuynzereb was this change requested in a issue?

Nope, it was not

**Comment by jeremyjordan** [comment]
@kuynzereb you mentioned:
>All callbacks calls are unified and take no additional arguments. That is, from now on it is on_epoch_end() instead of on_epoch_end(epoch, logs). Instead of these additional arguments now callback will have a link to the trainer, so it will have access to the current_epoch, global_step, callback_metrics and so on.

I'm wondering, was this `on_batch_start` intended to still have a batch argument?

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L462

**Comment by kuynzereb** [comment]
@jeremyjordan, well, it is not a callback, but a model hook. And actually I didn't think about them while doing this PR. But yes, it seems that the concept of model hooks is very similar to the concept of callbacks. So maybe it should be unified as well.

**Comment by jeremyjordan** [comment]
ah yes, that's right. i was going through `training_loop.py` quickly searching for `on_*` methods to fix a merge conflict and didn't stop to think about the difference between the two.
